include/llvm-dialects/Dialect/OpSet.h

```diff
@@ -0,0 +1,227 @@
+
+    if (Kind != OpDescription::Kind::Dialect && !hasOverload)
+      return false;
+
+    return desc.getMnemonic() == mnemonic && hasOverload == isOverload;
+  }
+};
+
+// An OpSet defines a set of operations. It is used to simplify operating on a
+// set of dialect operations, for instance, in the Visitor.
+class OpSet final {
+public:
+  // -------------------------------------------------------------
+  // Convenience functions to generate an OpSet from a given range
+  // of operations.
+  // -------------------------------------------------------------
+
+  // Construct an OpSet from a set of core opcodes.
+  static OpSet fromCoreOpcodes(ArrayRef<unsigned> ops) {
+    OpSet set;
+    for (const unsigned op : ops)
+      set.m_coreOpcodes.insert(op);
+
+    return set;
+  }
+
+  // Construct an OpSet from a set of intrinsics.
+  static OpSet fromIntrinsicIDs(ArrayRef<unsigned> intrinsicIDs) {
+    OpSet set;
+    for (const unsigned intrinsicID : intrinsicIDs)
+      set.m_intrinsicIDs.insert(intrinsicID);
+
+    return set;
+  }
+
+  // Construct an OpSet from a set of OpDescriptions.
+  static OpSet fromOpDescriptions(ArrayRef<OpDescription> descs) {
+    OpSet set;
+    for (const OpDescription &desc : descs)
+      set.tryInsertOp(desc);
+
+    return set;
+  }
+
+  // Construct an OpSet from a set of dialect ops, given as template
+  // arguments.
+  template <typename... OpTs> static const OpSet get() {
+    static OpSet set;
+    (... && appendT<OpTs>(set));
+    return set;
+  }
+
+  // -------------------------------------------------------------
+  // contains check to check if a given operation is stored in the OpSet.
+  // -------------------------------------------------------------
+  // Checks if a given core opcode is stored in the set.
+  bool containsCoreOpcode(unsigned coreOpcode) const {
+    return m_coreOpcodes.contains(coreOpcode);
+  }
+
+  // Checks if a given intrinsic ID is stored in the set.
+  bool containsIntrinsicID(unsigned intrinsicID) const {
+    return m_intrinsicIDs.contains(intrinsicID);
+  }
+
+  // Checks if a given dialect operation is stored in the set.
+  template <typename OpT> bool contains() const {
+    static OpDescription desc = OpDescription::get<OpT>();
+    return contains(desc);
+  }
+
+  // Checks if a given OpDescription is stored in the set.
+  bool contains(const OpDescription &desc) const {
+    if (desc.isCoreOp()) {
+      for (const unsigned op : desc.getOpcodes()) {
+        if (containsCoreOpcode(op))
+          return true;
+      }
+    }
+
+    if (desc.isIntrinsic()) {
+      for (const unsigned op : desc.getOpcodes()) {
+        if (containsIntrinsicID(op))
+          return true;
+      }
+    }
+
+    return isMatchingDialectOp(desc.getMnemonic());
+  }
+
+  // Checks if `inst` belongs to the OpSet.
+  bool contains(const Instruction &inst) const {
+    for (const unsigned CoreOp : m_coreOpcodes) {
+      if (inst.getOpcode() == CoreOp)
+        return true;
+    }
```

Comment:
For performance, this should make use of the fact that `m_coreOpcodes` is a set ;)